In [1]:
f = open("memory_profiler.log", "r")
lines = [line.replace('\x00','').replace('\n','') for line in f.readlines()]
for i,line in enumerate(lines):
    if line.find('FORWARD PASS START') !=-1: start_index = i
    if line.find('FORWARD PASS END') !=-1: end_index = i
lines = lines[start_index+1:end_index]

In [6]:
memory_list = []
for line in lines:
    line = line[line.find('MiB')+3:]
    line = line[:line.find('MiB')]
    memory_list.append(float(line.strip()))
print(memory_list)

[49.9, 49.4, 0.1, 49.2, 49.0, 49.0, 98.2, 98.0, 98.0, 24.6, 24.5, 24.5, 49.2, 49.0, 49.0, 49.1, 49.0, 49.0, 49.3, 49.0, 49.0, 12.4, 12.3, 12.3, 24.8, 24.5, 24.5, 24.6, 24.5, 24.5, 24.9, 24.5, 24.5, 6.2, 6.1, 6.1, 12.8, 12.3, 12.3, 12.4, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 3.1, 3.1, 3.1, 6.3, 6.1, 6.1, 6.2, 6.1, 6.1, 6.2, 6.1, 6.1, 0.1]


In [7]:
layer_list = []
for line in lines:
    line = line[line.find('MiB')+3:]
    layer_list.append(line.split('=')[1])
print(layer_list)

[' self.conv1(a)', ' self.b1(out)', ' self.relu1(out)', ' self.conv20(out)', ' self.b20(out)', ' self.relu20(out)', ' self.conv21(out)', ' self.b21(out)', ' self.relu21(out)', ' self.conv30(out)', ' self.b30(out)', ' self.relu30(out)', ' self.conv31(out)', ' self.b31(out)', ' self.relu31(out)', ' self.conv40(out)', ' self.b40(out)', ' self.relu40(out)', ' self.conv41(out)', ' self.b41(out)', ' self.relu41(out)', ' self.conv50(out)', ' self.b50(out)', ' self.relu50(out)', ' self.conv51(out)', ' self.b51(out)', ' self.relu51(out)', ' self.conv60(out)', ' self.b60(out)', ' self.relu60(out)', ' self.conv61(out)', ' self.b61(out)', ' self.relu61(out)', ' self.conv70(out)', ' self.b70(out)', ' self.relu70(out)', ' self.conv71(out)', ' self.b71(out)', ' self.relu71(out)', ' self.conv80(out)', ' self.b80(out)', ' self.relu80(out)', ' self.conv81(out)', ' self.b81(out)', ' self.relu81(out)', ' self.conv90(out)', ' self.b90(out)', ' self.relu90(out)', ' self.conv91(out)', ' self.b91(out)', ' sel

### Note: The output for the above two cells for your Neural Network should be as shown below
### print(memory_list)
\[49.9, 49.4, 0.1, 49.2, 49.0, 49.0, 98.2, 98.0, 98.0, 24.6, 24.5, 24.5, 49.2, 49.0, 49.0, 49.1, 49.0, 49.0, 49.3, 49.0, 49.0, 12.4, 12.3, 12.3, 24.8, 24.5, 24.5, 24.6, 24.5, 24.5, 24.9, 24.5, 24.5, 6.2, 6.1, 6.1, 12.8, 12.3, 12.3, 12.4, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 12.3, 3.1, 3.1, 3.1, 6.3, 6.1, 6.1, 6.2, 6.1, 6.1, 6.2, 6.1, 6.1, 0.1\]
### print(layer_list)
\[' self.conv1(a)', ' self.b1(out)', ' self.relu1(out)', ' self.conv20(out)', ' self.b20(out)', ' self.relu20(out)', ' self.conv21(out)', ' self.b21(out)', ' self.relu21(out)', ' self.conv30(out)', ' self.b30(out)', ' self.relu30(out)', ' self.conv31(out)', ' self.b31(out)', ' self.relu31(out)', ' self.conv40(out)', ' self.b40(out)', ' self.relu40(out)', ' self.conv41(out)', ' self.b41(out)', ' self.relu41(out)', ' self.conv50(out)', ' self.b50(out)', ' self.relu50(out)', ' self.conv51(out)', ' self.b51(out)', ' self.relu51(out)', ' self.conv60(out)', ' self.b60(out)', ' self.relu60(out)', ' self.conv61(out)', ' self.b61(out)', ' self.relu61(out)', ' self.conv70(out)', ' self.b70(out)', ' self.relu70(out)', ' self.conv71(out)', ' self.b71(out)', ' self.relu71(out)', ' self.conv80(out)', ' self.b80(out)', ' self.relu80(out)', ' self.conv81(out)', ' self.b81(out)', ' self.relu81(out)', ' self.conv90(out)', ' self.b90(out)', ' self.relu90(out)', ' self.conv91(out)', ' self.b91(out)', ' self.relu91(out)', ' self.conv100(out)', ' self.b100(out)', ' self.relu100(out)', ' self.conv101(out)', ' self.b101(out)', ' self.relu101(out)', ' self.conv110(out)', ' self.b110(out)', ' self.relu110(out)', ' self.conv111(out)', ' self.b111(out)', ' self.relu111(out)', ' self.conv120(out)', ' self.b120(out)', ' self.relu120(out)', ' self.conv121(out)', ' self.b121(out)', ' self.relu121(out)', ' self.conv130(out)', ' self.b130(out)', ' self.relu130(out)', ' self.conv131(out)', ' self.b131(out)', ' self.relu131(out)', ' self.conv140(out)', ' self.b140(out)', ' self.relu140(out)', ' self.conv141(out)', ' self.b141(out)', ' self.relu141(out)', ' self.avgpl(out)'\]
### If the file was not parsed properly or the output is not as expected please compile the memory_list and layer_list manually using the memory_profiler provided in Step1 or using any other technique.

In [10]:
### Algorithm to calculate the optimal Checkpointing Configuration using empirical method based on the memory profiled in previous step.

def check(mid, array, n, K): 
    count = 0
    sum = 0
    initial_pass_memory = 0
    checkpoint_indices = []
    for i in range(n): 
          
        # If individual element is greater maximum possible sum 
        if (array[i] > mid): 
            return False, initial_pass_memory, checkpoint_indices
  
        # Increase sum of current sub - array 
        sum += array[i] 
  
        # If the sum is greater than mid increase count 
        if (sum > mid): 
            count += 1
            checkpoint_indices.append(i)
            sum = 0
            initial_pass_memory += array[i]
    count += 1
  
    # Check condition 
    if (count <= K): return True, initial_pass_memory, checkpoint_indices
    return False, initial_pass_memory, checkpoint_indices
  
# Function to find maximum subarray sum which is minimum 
def solve(array, n, K): 
    start = 1
    end = 0
  
    for i in range(n): 
        end += array[i] 
  
    # Answer stores possible maximum sub array sum 
    answer = 0
    optimal_initial_pass_memory = 0
    optimal_checkpoint_indices = []
    while (start <= end): 
        mid = (start + end) // 2
  
        # If mid is possible solution Put answer = mid; 
        boolean, initial_pass_memory, checkpoint_indices = check(mid, array, n, K)
        if boolean: 
            answer = mid 
            optimal_initial_pass_memory = initial_pass_memory
            optimal_checkpoint_indices = checkpoint_indices
            end = mid - 1
        else: 
            start = mid + 1
  
    return answer, optimal_initial_pass_memory, checkpoint_indices

def get_optimal_checkpoint_config(memory_list, verbose=False):

    array = memory_list
    n = len(array)
    min_memory = -1
    min_config = None

    ### NOTE: n_checkpoints is the maximum no. of checkpoints that can be used
    for n_checkpoints in range(1,n):
        max_segment_memory, optimal_initial_pass_memory, optimal_checkpoint_indices = solve(array, n, n_checkpoints)
        total_memory = max_segment_memory + optimal_initial_pass_memory
        if (total_memory==0) or (max_segment_memory==0):continue
        if (min_memory==-1) or (total_memory<min_memory):
            min_memory=total_memory
            min_config = {  'max_segment_memory':max_segment_memory,
                            'optimal_initial_pass_memory':optimal_initial_pass_memory,
                            'total_memory':total_memory,
                            'n_checkpoints':n_checkpoints,
                            'optimal_checkpoint_indices':optimal_checkpoint_indices }

        if verbose:
            print('\nn_checkpoints =',n_checkpoints)
            print('max_segment_memory + optimal_initial_pass_memory =',round(max_segment_memory,1),'+',round(optimal_initial_pass_memory,1),'=',round(max_segment_memory+optimal_initial_pass_memory,1))
            print('Checkpoint indices=',optimal_checkpoint_indices)

    return min_config

min_config = get_optimal_checkpoint_config(memory_list, verbose=True)


n_checkpoints = 2
max_segment_memory + optimal_initial_pass_memory = 896.0 + 49.0 = 945.0
Checkpoint indices= [17, 80]

n_checkpoints = 3
max_segment_memory + optimal_initial_pass_memory = 590.0 + 49.0 = 639.0
Checkpoint indices= [11, 28]

n_checkpoints = 4
max_segment_memory + optimal_initial_pass_memory = 443.0 + 159.3 = 602.3
Checkpoint indices= [7, 17, 37, 80]

n_checkpoints = 5
max_segment_memory + optimal_initial_pass_memory = 345.0 + 183.9 = 528.9
Checkpoint indices= [6, 13, 21, 42, 74]

n_checkpoints = 6
max_segment_memory + optimal_initial_pass_memory = 278.0 + 233.5 = 511.5
Checkpoint indices= [6, 12, 18, 29, 51]

n_checkpoints = 7
max_segment_memory + optimal_initial_pass_memory = 241.0 + 281.9 = 522.9
Checkpoint indices= [5, 8, 15, 20, 32, 54]

n_checkpoints = 8
max_segment_memory + optimal_initial_pass_memory = 198.0 + 294.1 = 492.1
Checkpoint indices= [5, 8, 14, 19, 28, 42, 59]

n_checkpoints = 9
max_segment_memory + optimal_initial_pass_memory = 185.0 + 343.7 = 528.7
Ch

In [11]:
print('Checkpointing Configuration:')
print(min_config)

Checkpointing Configuration:
{'max_segment_memory': 155.0, 'optimal_initial_pass_memory': 324.80000000000007, 'total_memory': 479.80000000000007, 'n_checkpoints': 10, 'optimal_checkpoint_indices': [4, 7, 11, 15, 19, 26, 34, 48, 61]}


In [23]:
# Print the Layers in each segment for the optimal checkpointing configuration
checkpoints = min_config['optimal_checkpoint_indices']
# checkpoints = [5, 8, 14, 19, 28, 42, 59] # Enter manually here for any other configuration corresponding to different n_checkpoints

start_index = 0
for i, checkpoint in enumerate(checkpoints):
    end_index = checkpoint
    print('Segment:',i+1)
    for layer, memory in zip(layer_list[start_index:end_index],memory_list[start_index:end_index]):
        print('Layer =',layer,', Memory =',memory)
    print('\n')
    start_index = end_index

# Last Segment
print('Segment:',len(checkpoints)+1)
for layer, memory in zip(layer_list[start_index:],memory_list[start_index:]):
    print('Layer =',layer,', Memory =',memory)
print('\n')

Segment: 1
Layer =  self.conv1(a) , Memory = 49.9
Layer =  self.b1(out) , Memory = 49.4
Layer =  self.relu1(out) , Memory = 0.1
Layer =  self.conv20(out) , Memory = 49.2


Segment: 2
Layer =  self.b20(out) , Memory = 49.0
Layer =  self.relu20(out) , Memory = 49.0
Layer =  self.conv21(out) , Memory = 98.2


Segment: 3
Layer =  self.b21(out) , Memory = 98.0
Layer =  self.relu21(out) , Memory = 98.0
Layer =  self.conv30(out) , Memory = 24.6
Layer =  self.b30(out) , Memory = 24.5


Segment: 4
Layer =  self.relu30(out) , Memory = 24.5
Layer =  self.conv31(out) , Memory = 49.2
Layer =  self.b31(out) , Memory = 49.0
Layer =  self.relu31(out) , Memory = 49.0


Segment: 5
Layer =  self.conv40(out) , Memory = 49.1
Layer =  self.b40(out) , Memory = 49.0
Layer =  self.relu40(out) , Memory = 49.0
Layer =  self.conv41(out) , Memory = 49.3


Segment: 6
Layer =  self.b41(out) , Memory = 49.0
Layer =  self.relu41(out) , Memory = 49.0
Layer =  self.conv50(out) , Memory = 12.4
Layer =  self.b50(out) , Me

In [ ]:
# NOTE: Distribute the layers of the Neural Network as above among different segments in Step3_Checkpointing_Neural_Network.py.